[10.0][MIG] auth_jwt from 14.0

sbidoul opened the pull request.

sbidoul committed:
[ADD] auth_jwt

sbidoul committed:
auth_jwt: use PyJWT instead of python-jose

Because it allows validating with a list of audiences.

sbidoul committed:
auth_jwt: add signature algorithms

sbidoul committed:
auth_jwt: support multiple audiences

sbidoul committed:
auth_jwt: add nbf validation test

sbidoul committed:
auth_jwt: docs clarification and fixes

sbidoul committed:
auth_jwt: fix jwks URI support

Make it work with pyjwt.

sbidoul committed:
[MIG] auth_jwt

sbidoul committed:
[MIG] auth_jwt

sbidoul committed:
auth_jwt: mock instead of committing in tests

sbidoul committed:
auth_jwt: more precise precondition check

sbidoul committed:
Rename auth_jwt_test to auth_jwt_demo

sbidoul committed:
Rename auth_jwt_test to auth_jwt_demo

oca-travis committed:
[UPD] Update auth_jwt.pot

oca-travis committed:
[UPD] Update auth_jwt_demo.pot

OCA-git-bot committed:
[UPD] README.rst

OCA-git-bot committed:
[UPD] README.rst

OCA-git-bot committed:
auth_jwt 14.0.1.0.1

OCA-git-bot committed:
auth_jwt_demo 14.0.1.0.1

sbidoul committed:
[MIG] auth_jwt: change version

sbidoul committed:
[MIG] auth_jwt: setup.py

sbidoul committed:
[MIG] auth_jwt: no f-strings in python2

sbidoul committed:
[MIG] auth_jwt: python2 super()

sbidoul committed:
[MIG] auth_jwt: python2 mock

sbidoul committed:
[MIG] auth_jwt: python2 is_identifier

sbidoul committed:
[MIG] auth_jwt: use jose instead of pywjt

jose had a recent release with python 2 support,
while pyjwt's latest release with python 2 support
was from 2018.

Since jose does not support it, we don't support a validator with multiple audiences.

sbidoul committed:
[MIG] auth_jwt_test: change version

sbidoul committed:
[MIG] auth_jwt_demo: setup.py

sbidoul committed:
[MIG] auth_jwt_demo: save_session not supported in v10

sbidoul committed:
[MIG] auth_jwt_demo: HttpCase.url_open does not support headers

sbidoul committed:
[MIG] auth_jwt: _authenticate signature changed

sbidoul committed:
[MIG] auth_jwt_demo: disable test that is not working in Odoo 10

sbidoul committed:
auth_jwt: add public_or_jwt auth method

This method is useful for public endpoints that need
to work for anonymous user, but can be enhanced when
an authenticated user is know.

A typical use case is a "add to cart" enpoint that can
work for anonymous users, but can be enhanced by
binding the cart to a known customer when the authenticated
user is known.

sbidoul commented:
@lmignon @len-foss I have backported #294 (public_or_auth_jwt) here.

lmignon commented:
@sbidoul Thank you!

lmignon commented on auth_jwt/models/auth_jwt_validator.py:
```diff
@@ -187,12 +187,22 @@ def _register_auth_method(self):
             )
+            setattr(
+                IrHttp.__class__,
+                f"_auth_method_public_or_jwt_{rec.name}",
```
@sbidoul `f"...{}"` is not possible with PY2

lmignon committed:
[FIX] auth_jwt: Fix backport

lmignon committed:
FIX rsa dependency version for py2.7

lmignon committed:
[FIX] auth_jwt: Add missing UTF-8 coding comment

lmignon committed:
[FIX] auth_jwt:  handle cors preflight with custom auth

Odoo provides basic handling of CORS preflight requests: if an endpoint is marked as `cors=<truthy value>` then it'll automatically reply allowing the request.

*However* this is performed in `HttpRequest.dispatch` (likely in order to correctly handle the nodb case), which means it's executed after the auth handler has run... which means custom auth handlers will be called on preflight requests.

This change handle this preflight in the case of enpoint proteced by jwt auth

lmignon committed:
[FIX] auth_jwt: Allows requested headers into preflight request resonse

lmignon committed:
[FIX] auth_jwt: Add allowed methors into the preflight response header
